### DISCUSSION TITLE
pair plot with 2 variables looks different than >2

### DISCUSSION THREAD

**Discussion by jecampagne** [arviz-devs/arviz#2133] [Q&A]
Hello,
Here is a snippet
```python
!pip install matplotlib==3.5.0
import matplotlib as mpl
#print(mpl.__version__)   #verif 3.5.0 
import matplotlib.pyplot as plt
import arviz as az
#print(az.__version__) #verif 0.12.1
import numpy as np

def generator(dim):
  mean = [0]*dim
  A = np.random.rand(dim, dim)
  cov = np.dot(A, A.transpose())
  out = np.random.multivariate_normal(mean, cov, 5000).T
  data = {}
  for i in range(len(out)):
    data[f"var_{i}"] = out[i]
  return data

data = generator(dim=2)
axs= az.plot_pair(
            data,
            kind="kde",
            figsize=(5,5),
            marginal_kwargs={"plot_kwargs": {"c": "b"}},
            kde_kwargs={
                "hdi_probs": [0.68, 0.9],
                "contour_kwargs":{"colors":('r', 'green', 'blue'), "linewidths":3},
                "contourf_kwargs":{"alpha":0},
            },
            point_estimate_kwargs={"lw": 3, "c": "b"},
            marginals=True, textsize=10, point_estimate="median",
        );
```
gives
![image](https://user-images.githubusercontent.com/20539759/195034077-81263232-3cd6-46b7-aadd-67507ce9ec7e.png)

while with 
```python
data = generator(dim=3)
```
the `pair_plot`looks like
![image](https://user-images.githubusercontent.com/20539759/195034502-d0261ae6-9d23-4e72-bd63-0d9a76b8d283.png)

Q1: I would like to get the "dim=3" drawing of the marginals for the "dim=2" use case, is-it possible?
Q2: As a side remark, I do not manage to use az.convert_XYZ function to get something working for az.pair_plot it is wht I have used a home-made "data" generator. It would be nice if there exists a simple example to get it right that could be nice.

Thanks.

**Comment by OriolAbril** [comment]
It is currently not possible to avoid rotating the marginals when there are only two of them. It is hardcoded in https://github.com/arviz-devs/arviz/blob/main/arviz/plots/backends/matplotlib/pairplot.py#L162-L171 in case you'd be interested in sending a PR for this.

I believe I can help with the second question though. When using `az.convert_to_inference_data` or `az.from_dict`, ArviZ needs a convention in order to name each of the available dimensions. That convention is:

* 1D array -> `draw`
* 2D array -> `chain, draw`
* 3+D array -> `chain, draw, *shape`. The `*shape` part means taking as many extra dims as needed and naming them `{var_name}_dim_0`, `{var_name}_dim_1`... Moreover, if no var name is provided (aka, passing an array to convert_to_inference_data), `x` is used as variable name.

I imagine your code didn't work initially because you were giving a 2d array but it had `draw, dim_0` dimensions, not `chain, draw`. Converting that to a dict of 1D arrays fixed it because now `draw` matched the ArviZ convention. You could update your code to use a 3D array:

```
def generator(dim):
   mean = [0]*dim
   A = np.random.rand(dim, dim)
   cov = np.dot(A, A.transpose())
   return np.random.multivariate_normal(mean, cov, size=(1, 5000))
```

so when you do `az.convert_to_inference_data(data)` (or pass `data` straight to ArviZ functions) you will get:

```none
<xarray.Dataset>
Dimensions:  (chain: 1, draw: 5000, x_dim_0: 2)
Coordinates:
  * chain    (chain) int64 0
  * draw     (draw) int64 0 1 2 3 4 5 6 7 ... 4993 4994 4995 4996 4997 4998 4999
  * x_dim_0  (x_dim_0) int64 0 1
Data variables:
    x        (chain, draw, x_dim_0) float64 0.2053 0.1567 ... 1.253 1.416
Attributes:
    created_at:     2023-01-25T00:07:24.536681
    arviz_version:  0.14.0
```

and plots will work.

Personally, I would recommend defining the variable name yourself and converting explicitly, for example with `from_dict({"var": data})

**Comment by muhammadalmaskhan** [comment]
Hi @jecampagne,
Regarding your questions:

Getting the dim=3 style marginals for the dim=2 case:
The difference arises because dim=2 produces only a single pair of variables, so az.plot_pair generates marginals for just those two variables. To achieve a dim=3-like layout (with marginals along the diagonal for all variables) when dim=2, you can explicitly add a dummy variable to your dataset to create a consistent grid. For example:


``` data = generator(dim=2) 
data["var_dummy"] = np.zeros(5000)  # Add a dummy variable
az.plot_pair(
    data,
    kind="kde",
    figsize=(5, 5),
    marginal_kwargs={"plot_kwargs": {"c": "b"}},
    kde_kwargs={
        "hdi_probs": [0.68, 0.9],
        "contour_kwargs": {"colors": ('r', 'green', 'blue'), "linewidths": 3},
        "contourf_kwargs": {"alpha": 0},
    },
    point_estimate_kwargs={"lw": 3, "c": "b"},
    marginals=True,
    textsize=10,
    point_estimate="median", 
    ```
```
This will add a placeholder variable and mimic the layout seen in the dim=3 plot.

Using az.convert_to_inference_data:
The function az.convert_to_inference_data expects data in specific formats (e.g., numpy arrays, xarray datasets, or dictionaries). If you want to make your generator output compatible, try wrapping the dictionary in az.convert_to_inference_data:

```  data = generator(dim=2)
inference_data = az.convert_to_inference_data(data)
az.plot_pair(inference_data, ...) ```
```

Alternatively, you can skip using convert_to_inference_data and continue passing your dictionary directly to az.plot_pair, as you've done in your snippet.

Let me know if further clarification is needed!

~almas
